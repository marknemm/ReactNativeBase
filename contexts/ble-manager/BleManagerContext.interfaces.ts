import type BleManager from '@util/ble-manager';
import type { BleManagerContext } from './BleManagerContext';

/**
 * The {@link BleManagerContext} type.
 */
export interface TBleManagerContext {

  /**
   * The {@link BleManager} instance.
   */
  bleManager: BleManager;

  /**
   * Resets the {@link BleManager} instance.
   */
  resetBleManager: () => void;

}
