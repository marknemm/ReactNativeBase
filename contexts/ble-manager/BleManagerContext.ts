import type BleManager from '@util/ble-manager';
import { createContext } from 'react';
import type { TBleManagerContext } from './BleManagerContext.interfaces';

/**
 * A context that provides the {@link BleManager} instance.
 */
export const BleManagerContext: React.Context<TBleManagerContext> = createContext(null);

export type * from './BleManagerContext.interfaces';
