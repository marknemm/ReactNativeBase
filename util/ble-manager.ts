import { BleManager as NativeBleManager, ConnectionOptions, Device, ScanOptions, State, Subscription } from 'react-native-ble-plx';
import { logErr } from './log';
import { getPairedDeviceIds, isDevicePaired, pushPairedDevice, removePairedDevice } from './paired-device-storage';

export { Device, State };

/**
 * A proxy for the {@link NativeBleManager BleManager} class.
 */
export default class BleManager {

  #bleManager = new NativeBleManager();

  #connectedDevices = new Map<string, Device>();

  #connectedDevicesPromise: Promise<Map<string, Device>> = this.#reconnectDevices().catch((err) => {
    logErr(err);
    return this.#connectedDevices;
  });

  // Private setup methods //

  /**
   * Reconnects to all paired devices.
   *
   * @returns Promise which emits when all paired devices are reconnected.
   */
  async #reconnectDevices(): Promise<Map<string, Device>> {
    const pairedDeviceIds = getPairedDeviceIds();
    await this.waitForPoweredOnState();

    for (const deviceId of pairedDeviceIds) {
      try {
        await this.connectToDevice(deviceId);
      } catch (err) {
        logErr(err);
      }
    }

    return this.#connectedDevices;
  }

  // Public methods //

  /**
   * Gets the connected devices.
   *
   * @returns Promise which emits a map of connected devices.
   */
  async getConnectedDevices(): Promise<Map<string, Device>> {
    return this.#connectedDevicesPromise;
  }

  /**
   * Waits for the {@link State.PoweredOn PoweredOn} Bluetooth state to be reached.
   *
   * @returns Promise which emits when Bluetooth is powered on.
   */
  async waitForPoweredOnState(): Promise<void> {
    return new Promise((resolve) => {
      const subscription = this.onStateChange((state) => {
        if (state === State.PoweredOn) {
          subscription.remove();
          resolve();
        }
      });
    });
  }

  // Proxy methods //

  /**
   * Notifies about {@link State} changes of a {@link BleManager}.
   *
   * @param listener Callback which emits {@link State} changes of BLE Manager.
   * Also emits current state when registered.
   * @returns Subscription on which remove() function can be called to unsubscribe.
   */
  onStateChange(listener: (state: State) => void): Subscription {
    return this.#bleManager.onStateChange(listener, true);
  }

  /**
   * Current, global {@link State} of a {@link BleManager}. All APIs are working only when active state is `PoweredOn`.
   *
   * @returns Promise which emits current {@link State} of {@link BleManager}.
   */
  async state(): Promise<State> {
    return this.#bleManager.state();
  }

  /**
   * Starts device scanning. When previous scan is in progress it will be stopped before executing this command.
   *
   * @param listener Listener Function which will be called for every scanned Device (devices may be scanned multiple times).
   * @param UUIDs Array of strings containing UUIDs of Services which are registered in scanned Device. If null is passed, all available Devices will be scanned.
   * @param options Optional configuration for scanning operation.
   */
  startDeviceScan(
    listener: (scannedDevice: Device) => void,
    UUIDs: string[] = [],
    options: ScanOptions = { allowDuplicates: false }
  ) {
    this.#bleManager.startDeviceScan(UUIDs, options, async (err, device) => {
      if (err) {
        logErr(err);
        this.startDeviceScan(listener, UUIDs, options); // Must restart on error, will stop any previous scan
      } else if (device) {
        const isPaired = isDevicePaired(device.id);
        if (isPaired && !await device.isConnected()) {
          await this.getConnectedDevices(); // Ensure connected devices are up-to-date
        }
        listener(device);
      }
    });
  }

  /**
   * Stops {@link Device} scan if in progress.
   */
  stopDeviceScan() {
    this.#bleManager.stopDeviceScan();
  }

  /**
   * Connects to {@link Device} with provided ID.
   * Also, discovers all services and characteristics of the {@link Device}.
   *
   * @param deviceIdentifier {@link Device} identifier.
   * @param options Platform specific options for connection establishment.
   * @returns Promise that resolves to the connected {@link Device} object if successful.
   */
  async connectToDevice(deviceIdentifier: string, options: ConnectionOptions = {}): Promise<Device> {
    const device = await this.#bleManager.connectToDevice(deviceIdentifier, options);
    if (device) {
      this.#connectedDevices.set(deviceIdentifier, device);
      pushPairedDevice(deviceIdentifier);
      await device.discoverAllServicesAndCharacteristics();
    }
    return device;
  }

  /**
   * Disconnects from {@link Device} if it's connected or cancels pending connection.
   *
   * @param deviceIdentifier {@link Device} identifier to be closed.
   * @returns Promise that resolves to the closed {@link Device} when operation is successful.
   */
  async disconnectFromDevice(deviceIdentifier: string): Promise<Device> {
    this.#connectedDevices.delete(deviceIdentifier);
    removePairedDevice(deviceIdentifier);
    return this.#bleManager.cancelDeviceConnection(deviceIdentifier);
  }

  /**
   * Destroys the {@link BleManager} instance.
   * All operations which were in progress completes with `#bleerrorcodebluetoothmanagerdestroyed`|`BluetoothManagerDestroyed` error code.
   */
  destroy() {
    this.#bleManager.destroy();
    this.#connectedDevices.clear();
  }

}
