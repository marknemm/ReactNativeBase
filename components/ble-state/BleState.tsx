import StatusDot from '@components/status-dot/StatusDot';
import Tooltip from '@components/tooltip/Tooltip';
import { useBleState } from '@hooks/ble-hooks';
import { Text } from '@rneui/themed';
import { State } from '@util/ble-manager';
import { View } from 'react-native';
import { useStyles } from './BleState.styles';

/**
 * A component for displaying the BLE {@link State}.
 *
 * @returns The {@link BleState} component.
 */
const BleState: React.FC = () => {
  const styles = useStyles();
  const bleState = useBleState();
  const bleStatusColor = bleState === State.PoweredOn
    ? 'green'
    : 'red';

  return (
    <View style={styles.container}>
      <Tooltip popover={<Text style={styles.tooltipText}>{ bleState }</Text>}>
        <View style={styles.container}>
          <StatusDot color={bleStatusColor} />
          <Text>BLE Status</Text>
        </View>
      </Tooltip>
    </View>
  );
};

export default BleState;
