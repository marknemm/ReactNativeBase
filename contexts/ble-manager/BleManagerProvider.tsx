import BleManager from '@util/ble-manager';
import { useEffect, useMemo, useState, type PropsWithChildren } from 'react';
import { BleManagerContext } from './BleManagerContext';

/**
 * Provides the {@link BleManager} instance to the application.
 *
 * @param props The component {@link PropsWithChildren Props}.
 * @returns The {@link BleManagerProvider} component.
 */
const BleManagerProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [bleManager, setBleManager] = useState(() => new BleManager());

  // Destroy the BleManager instance when the component is unmounted.
  // Will call destroy on initial instance, but that will affect the current instance.
  useEffect(() => () => bleManager.destroy(), []); // eslint-disable-line react-hooks/exhaustive-deps

  const bleManagerCtx = useMemo(() => ({
    bleManager,
    resetBleManager: () => {
      bleManager.destroy();
      setBleManager(new BleManager());
    },
  }), [bleManager]);

  return (
    <BleManagerContext.Provider value={bleManagerCtx}>
      { children }
    </BleManagerContext.Provider>
  );
};

export type * from './BleManagerContext.interfaces';
export default BleManagerProvider;
