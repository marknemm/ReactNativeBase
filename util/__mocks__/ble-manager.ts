import { BleManagerContext } from '@contexts/ble-manager/BleManagerContext';
import AppProvider from '@test/contexts/app/AppProvider';
import { stubAllMethods } from '@test/util/spy';
import { renderHook } from '@testing-library/react-native';
import { useContext } from 'react';
import { Device, State, type NativeDevice } from 'react-native-ble-plx';

/**
 * Generates a map of default mock BLE devices.
 *
 * @returns A map of default mock BLE devices.
 */
export function genMockBleDevices(): Map<string, Device> {
  return new Map<string, Device>([
    ['device-3', genMockBleDevice({ id: 'device-3', name: 'Device 3', connected: true })],
    ['device-2', genMockBleDevice({ id: 'device-2', name: 'Device 2' })],
    ['device-0', genMockBleDevice({ id: 'device-0', name: 'Device 0', localName: 'Device A' })],
    ['device-1', genMockBleDevice({ id: 'device-1', name: 'Device 1', localName: 'Device B' })],
    ['device-4', genMockBleDevice({ id: 'device-4' })],
  ]);
}

/**
 * Gets the singleton {@link BleManager} instance from the test {@link BleManagerContext}.
 *
 * @returns The singleton {@link BleManager} instance.
 */
export function getMockBleManager() {
  return renderHook(
    () => useContext(BleManagerContext),
    { wrapper: AppProvider }
  ).result.current.bleManager;
}

/**
 * The `@util/ble-manager` module mock.
 */
const BleManagerMock = jest.fn().mockImplementation(() => {
  let mockDevices = genMockBleDevices();

  return {
    getConnectedDevices: jest.fn().mockResolvedValue(mockDevices),
    waitForPoweredOnState: jest.fn().mockResolvedValue(null),
    onStateChange: jest.fn().mockImplementation((listenerCb) => {
      listenerCb?.(State.PoweredOn);
      return { remove: jest.fn() };
    }),
    state: jest.fn().mockResolvedValue(State.PoweredOn),
    startDeviceScan: jest.fn().mockImplementation((listenerCb) => {
      for (const device of mockDevices.values()) {
        listenerCb?.(device);
      }
    }),
    stopDeviceScan: jest.fn(),
    connectToDevice: jest.fn().mockImplementation((deviceIdentifier: string) => {
      const device = mockDevices.get(deviceIdentifier) ?? genMockBleDevice(deviceIdentifier);
      (device.isConnected as jest.Mock).mockResolvedValue(true);
      return Promise.resolve(device);
    }),
    disconnectFromDevice: jest.fn().mockImplementation((deviceIdentifier: string) => {
      const device = mockDevices.get(deviceIdentifier) ?? genMockBleDevice(deviceIdentifier);
      (device.isConnected as jest.Mock).mockResolvedValue(false);
      return Promise.resolve(device);
    }),
    destroy: jest.fn(),

    // Additional helper methods for manipulating the mock devices.
    mockDevices,
    resetMockDevices: () => { mockDevices = genMockBleDevices(); },
  };
});

/**
 * Generates a mock {@link Device} object.
 *
 * @param data The {@link NativeDevice} data to populate the mock {@link Device}.
 * @returns A mock {@link Device} object.
 */
export function genMockBleDevice(
  data: (Partial<NativeDevice> & { connected?: boolean }) | string | number = 0
): Device {
  const device = new Device(null, null);

  if (typeof data === 'number') {
    data = { id: `device-${data}` };
  } else if (typeof data === 'string') {
    data = { id: data };
  }

  for (const key of Object.keys(data)) {
    (device as any)[key as keyof Device] = data[key as keyof NativeDevice];
  }
  stubAllMethods(device);
  (device.isConnected as jest.Mock).mockResolvedValue(data.connected ?? false);

  return device;
}

export { Device, State };
export default BleManagerMock;
