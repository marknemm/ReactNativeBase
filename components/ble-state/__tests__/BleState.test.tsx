import BleState from '@components/ble-state/BleState';
import { BleManagerContext } from '@contexts/ble-manager/BleManagerContext';
import AppProvider from '@test/contexts/app/AppProvider';
import { render, renderHook, screen } from '@testing-library/react-native';
import { State } from '@util/ble-manager';
import { useContext } from 'react';

describe('<BleState />', () => {
  describe('snapshots', () => {
    it('renders correctly when powered on', () => {
      render(
        <BleState />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });

    it('renders correctly when powered off', () => {
      const { bleManager } = renderHook(
        () => useContext(BleManagerContext),
        { wrapper: AppProvider }
      ).result.current;
      (bleManager.onStateChange as jest.Mock).mockImplementationOnce((cb) => cb(State.PoweredOff));

      render(
        <BleState />,
        { wrapper: AppProvider }
      );

      expect(screen.toJSON()).toMatchSnapshot();
    });
  });
});
