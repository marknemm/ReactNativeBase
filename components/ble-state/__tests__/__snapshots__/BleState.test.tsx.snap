// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BleState /> snapshots renders correctly when powered off 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "columnGap": 5,
      "flexDirection": "row",
      "flexGrow": 1,
    }
  }
>
  <View
    collapsable={false}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "columnGap": 5,
            "flexDirection": "row",
            "flexGrow": 1,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "red",
              "borderRadius": 5,
              "height": 10,
              "width": 10,
            }
          }
          testID="rnb-status-dot"
        />
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontFamily": undefined,
              "fontSize": 23.57142857142857,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "marginBottom": 0,
              "textAlign": "left",
              "textDecorationLine": "none",
              "textDecorationStyle": "solid",
            }
          }
        >
          BLE Status
        </Text>
      </View>
    </View>
  </View>
</View>
`;

exports[`<BleState /> snapshots renders correctly when powered on 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "columnGap": 5,
      "flexDirection": "row",
      "flexGrow": 1,
    }
  }
>
  <View
    collapsable={false}
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
    >
      <View
        style={
          {
            "alignItems": "center",
            "columnGap": 5,
            "flexDirection": "row",
            "flexGrow": 1,
          }
        }
      >
        <View
          style={
            {
              "backgroundColor": "green",
              "borderRadius": 5,
              "height": 10,
              "width": 10,
            }
          }
          testID="rnb-status-dot"
        />
        <Text
          accessibilityRole="text"
          style={
            {
              "color": "#242424",
              "fontFamily": undefined,
              "fontSize": 23.57142857142857,
              "fontStyle": "normal",
              "fontWeight": "normal",
              "marginBottom": 0,
              "textAlign": "left",
              "textDecorationLine": "none",
              "textDecorationStyle": "solid",
            }
          }
        >
          BLE Status
        </Text>
      </View>
    </View>
  </View>
</View>
`;
